### PR TITLE
Updated Python and PyMC, removed TensorFlow, and added PyTorch in conda environment.

### PR DESCRIPTION
**Pull request by samuelklee**
Copying over some discussion from Slack, with some slight modifications:

> I took a quick stab at updating the environment for gCNV. Even taking out TensorFlow (assuming that the CNN will not be supported by this environment), it's a difficult task:
> 1. The goal is to update Python from 3.6 to 3.10+, since Terra now requires the latter for officially supported images.
> 2. However, gCNV relies on the PyMC3 package. PyMC3 3.1 is currently used in GATK master. 3.1 was released in 2017, not long before our release of gCNV in 2018, but it's very old now.
> 3. The latest version of Python that is supported by PyMC3 3.1 in conda is Python 3.6.
> 4. @asmirnov239 has a draft PR (#8094) that updates PyMC3 to 3.5 and Python to 3.7, which clearly still falls short of Python 3.10+. This PR also updated some gCNV code to make it compatible with PyMC3 3.5. (It also removed TensorFlow and added PyTorch.)
> 5. @asmirnov239 also merged a PR that added tests for numerical reproducibility of GermlineCNVCaller in cohort mode in #7889.
> 6. The earliest version of PyMC that supports Python 3.10+ is PyMC 4, released in 2022.
> 7. However, PyMC 4 introduces API changes, which will also require additional gCNV code changes and numerical testing.
> 8. These API changes are because the underlying computational backend for PyMC was updated from Theano (think of this as an old alternative to TensorFlow) to Aesara.
> 9. Since then, PyMC 5.9 has been released and the underlying backend has been updated again, from Aesara to PyTensor.
> 10. So if we are going to update the environment to support Python 3.10+, it probably makes sense to go all the way to PyMC 5.9.

I've made some strides in this PR; as of [6b08f3a](https://github.com/broadinstitute/gatk/pull/8561/commits/6b08f3af205cb9af1f5c63a0786f9a5a52cd78c1), I've made enough updates to accommodate API changes so that cohort-mode inference for both GermlineCNVCaller and DetermineGermlineContigPloidy runs successfully under Python 3.10 and PyMC 5.9.0---although note that 5.9.1 has been released in the interim!

However, our work has just begun. Results now produced in the numerical tests mentioned above are quite far off from the original expected results. It remains to be seen whether this is due to the randomness of inference, some slight changes to the model prior that were necessitated by the API changes, or some bugs introduced in other code updates. (Also note that I believe Andrey's PR in item 4 already broke these tests, although the numerical differences were much smaller and more reasonable---but perhaps he can confirm. Also noting here that I think determinism is still currently broken as of this commit---there have been some changes to PyTensor/PyMC seeding so that our previous theano/PyMC3 hack no longer applies.)

So I think the next step is to just go to scientific-level testing and see what the fallout is. Ideally, we'd still get good performance (or perhaps better! at least on the runtime side, hopefully...) and we can just update the numerical tests. But if performance tanks, then we might need to see whether I've introduced any bugs. @mwalker174 @asmirnov239 perhaps you can comment on what might be the appropriate test suite here----1kGP?

I'll also highlight again that this PR will remove TensorFlow and might require that the corresponding CNN implementations be supported by an alternate strategy, at least until the PyTorch implementation goes in.

### CODE REVIEW

**Comment by mwalker174** [comment]
Thanks for your work on this @samuelklee! Testing on both wes and wgs would be ideal. For wgs we can use the gatk-sv reference panel, which is our standard (I can help with this once a docker is ready). For wes, 1kgp would work although it's definitely showing its age. Are the integration test differences large?

**Comment by samuelklee** [comment]
Gonna do some rebasing and splitting into commits, copying some bits from the current git log here for posterity:

```
commit 28174871c5b2e99a3044a30c0de24a63d2fee5bc (HEAD -> sl_python_version_update, origin/sl_python_version_update)
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 13 00:59:52 2023 -0500

    update gCNV WDL tests

commit 31a204b9e900849b5a313e161893de34a2094bb0
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 13 00:14:34 2023 -0500

    staged base rc1

commit 74f8fa724dfac142ccd7ac79a757c0e5ac3bb06c
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 13 00:01:38 2023 -0500

    minor pymc/pytensor version upgrades, fix 2-interval edge case, update some theano docs

commit 9c9d0c570dd2712631739e0a9d41e90c4ccd3456
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 23:36:55 2023 -0500

    update VETS expected, verbose conda env create, pin torch CPU MKL, add pysam, fixed more tests

commit c0a17dfcf9fa1139927570d2f16125bc15a2c19f
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 20:07:08 2023 -0500

    fix CNV plotting

commit dd2dd503a92e6fbb5a49be6a88d2e813eb8bf85b
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 15:14:08 2023 -0500

    update gCNV expected results, generated on WSL Ubuntu 20.04.2

commit 27d76e8f22d61df90eeb337e033ae128ce07ab90
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 14:53:04 2023 -0500

    update python env integration tests

commit 348df9192235f7d1ea941d0b31e5c96acc0d6491
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 10:59:23 2023 -0500

    disable CNN tests, add deprecation message

commit ed59372b4be226785af1d3fb1b1a39a9ad3b4f6a
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 12 09:55:24 2023 -0500

    clean up rebase

commit 18e530db26f803ee46a0006843cb36d4ed4194b4
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 11:31:46 2023 -0500

    postprocess fixed

commit f510c2e9f10d7066c15f1835669d676964b8a4cb
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 10:13:01 2023 -0500

    fix deprecated np.int in optimizer

commit 939a032f356f2f8f67b5aae426fc427d1d1ea6c4
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:50:57 2023 -0500

    remove unnecessary seeding in cohort denoising script

commit cf82ea5c99250f1784f8b1a9279e7dbb8841fa89
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:38:08 2023 -0500

    add back setup.py files

commit 8348f546de6b3d32e1f02f6851730226c0dbffc9
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:37:09 2023 -0500

    update pymc version in init

commit 850d60ef95b6126c05af9cd7c2cb528a306e1224
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:32:42 2023 -0500

    added pip editable docs

commit 9c51b311442b0796ab1224213e83290caea0f93f
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:24:50 2023 -0500

    whitespace

commit d9b180385168fdd1ef55cee8a1069fc1f7928f38
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:24:10 2023 -0500

    update setup_gcnvkernel.py and pin pytensor

commit 7ccbd6da3d4afd1c987a66f6874bc4918495f943
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 09:20:49 2023 -0500

    update conda in base and python packages

commit 693a1f9de10ee9950000abb83ef598cde82e026b
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 08:55:53 2023 -0500

    clean up Mixture, sample seeding, safelog

commit 2b211d7ed7875c798020ceaeed865523f25c7096
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Fri Dec 8 00:28:54 2023 -0500

    fixed all determinism, need to clean up seeds

commit 799228dd5fafa2bf5d57e65e9aab256cdfc4698a
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 7 22:22:18 2023 -0500

    more dCR, Mixture

commit 124073d9af37c19573f90270c3cb0f4ae5ba4dd0
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 7 19:39:55 2023 -0500

    fix dCR sampling?

commit f6871c9f12dbd52d84e78bba4f03d276ba1efb72
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 7 15:34:26 2023 -0500

    logsumexp cleanup

commit 0c6cba790d8c3566a1a872d292e2b7acc692670e
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Dec 7 13:57:16 2023 -0500

    revert backport of MeanField

commit 67fb373687d32ec7e0fa329d4f3864e2cccabc53
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 10:45:11 2023 -0500

    just used sample_node, finally deterministic?!

commit 754c424566cb9c191b9775b02d464488d7f68f68
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Dec 6 07:22:51 2023 -0500

    port stable logsumexp, though it doesn't seem to make a difference in ploidy

commit 95c944d792642ba5ec8dceaaff67d3a35cb3eab0
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 23:51:01 2023 -0500

    working with Mixture, still nondeterministic

commit 426375ac6473999ba249e775f2aa7d622534d510
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 22:32:44 2023 -0500

    stochastic node cleanup

commit dc66f3f6a07dc82ba4258b3c0a65d990355da8d7
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 21:38:52 2023 -0500

    safe log in log ploidy priors

commit 9712fa169a08edcf1a7a56622708e610b862631e
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Dec 5 21:35:05 2023 -0500

    still debugging stochastic node

commit a79762c616f8758e0fd073b9e8e5d1ad30c1d5d0
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Nov 9 16:48:05 2023 -0500

    blas in base, conda with libmamba solver

commit b4f5301c28a03689fca0b95ef652a51dd991686d
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Nov 9 12:45:53 2023 -0500

    freeze conda and set libmamba in base

commit f57a13a08fc3d049f0271e9aa94639ecb87b50f2
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Thu Nov 9 07:29:30 2023 -0500

    libmamba 23.9.0

commit 45058f27aae9c9240a167f126e32a6bddd3353ff
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Nov 8 23:33:51 2023 -0500

    conda 23.9.0

commit d95494d282dd42ea3377de6c2d3554a3a5db65e4
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 30 17:04:34 2023 -0400

    minor fixes to get ploidy working

commit c6a21f33f1f17b2bc9b878d8d6bfc6d7da784933
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 30 16:33:22 2023 -0400

    truncated normal fixes

commit 81e3fff0a099731730d1657fc12f56abdc2b2927
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 30 16:03:02 2023 -0400

    minor clip fix

commit 2fb94e88ab457ec037d304f7a80384b70a8685cc
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 30 15:17:06 2023 -0400

    use model.rvs_to_values for registry; inference completes but numerical tests seem off!

commit 5d2062f3301f682dbf427d898521ec6da8338944
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 30 15:03:40 2023 -0400

    VarMap and transformed name in register, exact match error to warn

commit 3dca890329396c9afae52aa8779cdfb97c18fa7c
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Wed Oct 25 23:28:32 2023 -0400

    inference works! oh wow

commit d056855608fdf222a38a25d8de41f11fa1e82c22
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Oct 24 22:11:30 2023 -0400

    past full denoising instantiation and through to first calling epoch before hitting another shape error

commit fcfdb29e79e0652d24a5f509ade6e941f84293c2
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Oct 24 00:17:46 2023 -0400

    revert pytensorf

commit 2ec79b362bdfcd34b3ca1736e6327ecbbe2269ad
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Tue Oct 24 00:16:04 2023 -0400

    HalfFlat typo

commit 6eb1594d62ed62b992db04c523ffc598845d3b4c
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 23 23:46:42 2023 -0400

    added TODOs for nontrivial changes

commit c01754082991c006c8d56f0027138948d792ce75
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 23 23:33:43 2023 -0400

    warmup denoising inference runs, a minor miracle! probably incorrect after some nontrivial changes, though

commit 3bf1c5b65d12fe76e499d1b43e58736e31f25f33
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 23 17:13:04 2023 -0400

    find and replace, some version bumps

commit 3039e1f01579003518200b9033c5674c3d1de156
Author: Samuel Lee <lee.samuel.k@gmail.com>
Date:   Mon Oct 23 14:30:32 2023 -0400

    python 3.10.9, pymc3 3.11.2

commit 4a2c86391990fb01d2e67fc7cf823495eef1ba99
Author: Andrey Smirnov <asmirnov2009@gmail.com>
Date:   Sun Oct 23 03:58:36 2022 +0000

    Updates to gCNV code for pymc3 upgrades.

```

**Comment by matthdsm** [comment]
Hi @samuelklee,

Any updates on this PR? Will this be able to get merged in the foreseeable future?

Thanks
M

**Comment by samuelklee** [comment]
Thanks for the inquiry, @matthdsm! This has indeed been a bit on the back burner since winter break, so perhaps it's time for a kick start.

@mwalker174 (or anyone else interested in running scientific tests), a Docker image is available at: us.gcr.io/broad-dsde-methods/broad-gatk-snapshots/gatk-remote-builds@sha256:59310a7e8d635d4c09a6b6e09d188070628a42f7110805eeb26ca044aa1f71a5 Note that I haven't yet used this to run any tests, so please let me know if you encounter any issues.

@droazen can you provide a roadmap for the CNN filtering tools? Those should be deprecated or otherwise managed before this PR is merged. I am not sure who the major stakeholders are here, but any users out there should feel free to chime in.

I will try to do a quick self-review and some minor cleanup.

My expectation is that the timescale for this to get merged will be on the order of a few weeks to months. But it would be great if we can get it in sooner!

**Comment by droazen** [comment]
@samuelklee My plan for the deprecation was to do it concurrently with (or just before) the merge of this PR. As this gets closer to a final merge, we can coordinate on that aspect.

**Comment by vdauwera** [comment]
Hi @samuelklee, @droazen, totally understand this is not high priority. Would you be able to estimate a rough timeframe when you might be able to tackle this? It would help our community figure out whether to wait or pursue other mitigation in the meantime.

**Comment by samuelklee** [comment]
Hi @vdauwera! I think we’re still within the “few weeks to months” timeframe quoted above, although we’re definitely pushing into the latter territory 😜 

Would be happy to chat more, if you like—would also be curious to hear what mitigation is being considered. Feel free to set up a meeting!

**Comment by droazen** [comment]
We are hoping to re-prioritize this and get it over the finish line soon! Thanks to @samuelklee 's heroic efforts we are 90% of the way up the path out of our self-imposed Python hell, but I believe there are still a few stray Python demons pulling at our ankles.

**Comment by samuelklee** [comment]
> Just noting here that @gokalpcelik was kind enough to run some tests on this branch and noticed that I probably broke posterior sampling—it appears that I reverted to the non-online, memory intensive implementation that we ran into previously during development. Need to carve out some time to fix this issue and more generally polish things up, but I have other more pressing deadlines at the moment.

@gokalpcelik I think I fixed this posterior-sampling issue. I haven't yet done my own testing, but let me know if you beat me to it!

**Comment by samuelklee** [comment]
Here's one, if anyone would like to try it out! us.gcr.io/broad-dsde-methods/broad-gatk-snapshots/gatk@sha256:c405d1d7fc42a008bb11a2f60f95bb65f1c487cf4986850a7d5d7d4bdfc37965

**Comment by samuelklee** [comment]
Finishing up a tieout using old MalariaGEN Pf7 CNV genotyping runs, and I think things look good from this perspective, at least.

This tieout uses a subset of the Pf7 samples containing 300 cohort and 1683 case samples (which were indeed treated as a cohort-case cluster in the original Pf7 CNV genotyping analysis). ~4k genomic bins are covered. We compare this branch against 4.5.0.0, as well as this branch against itself (checking for reproducibility).

Costs for this branch ($10.92) and 4.5.0.0 ($10.96) were quite comparable. Note that a small portion of these costs derives from Pf7-specific genotyping steps, which I did not bother to remove from the workflow.

Runtime for the ploidy modeling and postprocessing steps were comparable. Interestingly, **runtime for the gCNV was ~20-25% longer with this branch than with 4.5.0.0, but memory usage fell by a factor of ~3 (~6GB to ~2GB)!** I am not sure if we could recoup the runtime with some more tweaking of the environment (perhaps double checking that optimized BLAS/MKL/etc. packages are properly used, changing environment variables/flags, etc.), but I think the decrease in memory usage is quite nice.

Concordance was checked for the following quantities:

1) Variational posterior means (`mu_*`) and standard deviations (`std_*`) for all analogous variables in the ploidy and gCNV models. There were some slight changes to the gCNV model in this branch (e.g., the functional form of the ARD prior was changed), which means some variables are no longer directly comparable. Furthermore, some variables (such as the bias factors W) are degenerate and cannot be immediately compared. Otherwise, there is good concordance between the remaining variables, e.g.:

image
image
image
image
image
image

2) ... Will update more later!

**Comment by droazen** [comment]
Thanks for sharing these results @samuelklee ! That memory vs runtime tradeoff does not seem at all bad to me. 

What else needs to be done before we can feel confident about merging this (apart from deprecating the old CNN tools)? Is the posterior sampling issue resolved?

**Comment by samuelklee** [comment]
Yes, that's been resolved! Phew.

If anyone wants to start taking an initial review pass, that would be appreciated! There are probably a few very minor cleanups I could do (e.g., removing stray commented code and TODOs), but I don't think that should hold anything up.

**Comment by samuelklee** [comment]
Also, I think @mwalker174 should have the final say on merging, as he's the current gCNV owner. Would be great if he could run any GATK-SV-specific tests and report back on runtime/accuracy!

File: src/main/python/org/broadinstitute/hellbender/gcnvkernel/utils/math.py
Comment (samuelklee) [inline src/main/python/org/broadinstitute/hellbender/gcnvkernel/utils/math.py:81]: Whoops, probably should remove this!

File: src/test/java/org/broadinstitute/hellbender/utils/python/PythonEnvironmentIntegrationTest.java
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/utils/python/PythonEnvironmentIntegrationTest.java:45]: Is this a real TODO?

**Comment by droazen** [comment]
@samuelklee It's great to see this inching closer to finally getting merged! I think we should plan on merging it shortly after the upcoming 4.6 release, assuming the remaining testing doesn't reveal any issues. Since 4.6 will contain a number of important bug fixes and user-requested changes, I want to get it out before making a big breaking change to the Python environment. After 4.6 is out, we can meet briefly to discuss how best to handle the deprecation of the legacy CNN tools, and hopefully get this in at last.

**Review by ldgauthier — COMMENTED** [review]
Thanks for documenting all the API changes.  This looks reasonable to me (along with the malaria plots).  There are a couple more TODOs that I think are to-done.  @droazen is going to help us come up with a CNN deprecation/transition plan so we don't have to turn off tests to merge

**Comment by samuelklee** [comment]
@mwalker174 reported last week that exomes look good, with perhaps more details on cost/accuracy forthcoming!

I'll be OOO next week, but will address comments after I get back---already looked through them and I think it should be quick (famous last words).

**Comment by mwalker174** [comment]
I've run some basic WGS tests in gatk-sv, and the results look good. Using our 161-sample 30x NYGC 1KGP test panel, here is a summary of per-sample raw calls from gCNV+cnMOPs. That is, records look like:

```
chr1	103639835	103663835	1kgp_161_bwa_scramble_DEL_9527	HG01552	DEL	cnmops,gcnv
chr1	103643835	103645835	1kgp_161_bwa_scramble_DEL_9531	HG01494	DEL	gcnv
chr1	103643835	103652200	1kgp_161_bwa_scramble_DEL_9534	HG02236	DEL	cnmops,gcnv
chr1	103645600	103724500	1kgp_161_bwa_scramble_DEL_9538	HG03370	DEL	cnmops,gcnv
chr1	103647835	103649835	1kgp_161_bwa_scramble_DEL_9541	HG01495	DEL	gcnv
chr1	103647835	103681000	1kgp_161_bwa_scramble_DEL_9542	HG01494	DEL	cnmops,gcnv
…
```

Master:
DEL: 641700
DUP: 699063

Branch:
DEL: 640669
Master intersection: 635178 (99.14% sensitivity)
DUP: 691469
Master intersection: 677687 (98.00%)

Note that this is a simple bedtools intersection requiring 90% reciprocal overlap, but not matching samples.

Subsetting just to NA19420:

Master:
DEL: 4974
DUP: 4254

Branch:
DEL: 4958
Master intersection: 4797 (96.75%)
DUP: 4210
Master intersection: 3802 (90.30%)

Further subsetting to NA19240 variants over 5kbp, which is the default gatk-sv threshold for depth-only calls:

Master:
DEL: 1133
DUP: 916

Branch:
DEL: 1132
Master intersection: 1060 (93.64%)
DUP: 893
Master intersection: 757 (84.77%)

While it does appear there's an appreciable difference here, if we subset to NA19240 variants that survive gatk-sv filtering to the "CleanVcf" stage (in the current master), the differences are much less:

Master:
DEL: 225
DUP: 61

Branch:
DEL: 226
Master intersection: 223 (98.67%)
DUP: 58
Master intersection: 58 (100%)


I think the differences in raw calls is probably "in the noise" for WGS, given the high concordance in this test sample after gatk-sv filtering and genotyping are applied.

**Comment by mwalker174** [comment]
Also we've noticed a slight drop in peak memory usage from 8.7 GB to 7.2 (17% decrease) in both cohort and case modes for `GermlineCNVCaller` which is a welcome improvement!

**Comment by JMF47** [comment]
Here's the performance benchmarking
#      sen_del   sen_dup   ppv_del   ppv_dup sen_del_hist sen_dup_hist ppv_del_hist ppv_dup_hist
# 1  0.2000000 0.4121864 0.9387755 0.8521127          345          279           98          142
# 2  0.4598540 0.6086957 0.9647059 0.8602941          137          184           85          136
# 3  0.6352941 0.7181208 0.9565217 0.8650794           85          149           69          126
# 4  0.7796610 0.7744361 0.9636364 0.8606557           59          133           55          122
# 5  0.8913043 0.8715596 0.9583333 0.8495575           46          109           48          113
# 6  0.9024390 0.9019608 0.9750000 0.8545455           41          102           40          110
# 7  0.9428571 0.9263158 0.9722222 0.8666667           35           95           36          105
# 8  0.9310345 0.9534884 0.9666667 0.8736842           29           86           30           95
# 9  0.9285714 0.9518072 1.0000000 0.9213483           28           83           26           89
# 10 0.9565217 0.9487179 1.0000000 0.9259259           23           78           23           81

### NEW gCNV docker
#      sen_del   sen_dup   ppv_del   ppv_dup sen_del_hist sen_dup_hist ppv_del_hist ppv_dup_hist
# 1  0.2034884 0.4107143 0.9387755 0.8581560          344          280           98          141
# 2  0.4705882 0.6086957 0.9647059 0.8676471          136          184           85          136
# 3  0.6547619 0.7248322 0.9565217 0.8730159           84          149           69          126
# 4  0.7931034 0.7819549 0.9636364 0.8677686           58          133           55          121
# 5  0.9111111 0.8807339 0.9583333 0.8558559           45          109           48          111
# 6  0.9250000 0.9019608 0.9750000 0.8584906           40          102           40          106
# 7  0.9705882 0.9263158 0.9722222 0.8712871           34           95           36          101
# 8  0.9642857 0.9534884 0.9655172 0.8791209           28           86           29           91
# 9  0.9629630 0.9518072 1.0000000 0.9186047           27           83           25           86
# 10 1.0000000 0.9487179 1.0000000 0.9113924           22           78           22           79

**Comment by samuelklee** [comment]
Great, thanks @mwalker174 and @JMF47! I will try to get this tidied up before heading out for 7/4.

**Comment by matthdsm** [comment]
Hi all,

Any chance this will make it into a release soon? I was hoping this got merged with the recent docker image overhaul.

Thanks
Matthias

File: src/main/python/org/broadinstitute/hellbender/gcnvkernel/_version.py
Comment (samuelklee) [reply src/main/python/org/broadinstitute/hellbender/gcnvkernel/_version.py:1]: I don't believe so, but figured this merited a version bump!

File: src/main/python/org/broadinstitute/hellbender/gcnvkernel/inference/deterministic_annealing.py
Comment (samuelklee) [reply src/main/python/org/broadinstitute/hellbender/gcnvkernel/inference/deterministic_annealing.py:25]: Yes, this is actually present in Andrey's update to pymc 3.5; see https://github.com/broadinstitute/gatk/blob/d73809db2b358e3e52e89ecd45c90d84bb039829/src/main/python/org/broadinstitute/hellbender/gcnvkernel/inference/deterministic_annealing.py#L31.

File: src/main/python/org/broadinstitute/hellbender/gcnvkernel/models/pytensor_hmm.py
Comment (samuelklee) [reply src/main/python/org/broadinstitute/hellbender/gcnvkernel/models/pytensor_hmm.py:320]: Went through and cleaned up the TODOs I added!

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCallerIntegrationTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCallerIntegrationTest.java:164]: Good catch!

File: src/test/java/org/broadinstitute/hellbender/utils/python/PythonEnvironmentIntegrationTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/utils/python/PythonEnvironmentIntegrationTest.java:45]: Yes, TBH I'm not sure how to check for MKL versions now. But maybe it doesn't matter if runtime seems OK. I'll leave this here with an updated note just in case.

**Comment by samuelklee** [comment]
@matthdsm this was intentionally left out of the recent 4.6 release, but should go into the next minor release. Would of course appreciate any testing/feedback from the community before then!

File: .github/workflows/gatk-tests.yml
Comment (samuelklee) [reply .github/workflows/gatk-tests.yml:294]: @droazen leaving this as a reminder that we'll need to deal with the CNN tests/code/etc. disabled here in a subsequent PR.

**Comment by samuelklee** [comment]
Released gatkbase-3.3.0 to `broadinstitute/gatk:gatkbase-3.3.0`, but getting `Permission "artifactregistry.repositories.uploadArtifacts" denied on resource "projects/broad-gatk/locations/us/repositories/us.gcr.io"` when trying to push to `us.gcr.io/broad-gatk/gatk:gatkbase-3.3.0`.

**Comment by samuelklee** [comment]
Just added `@DeprecatedFeature` tags to the CNN tools. @droazen will help me push `broadinstitute/gatk:gatkbase-3.3.0` to `us.gcr.io/broad-gatk/gatk:gatkbase-3.3.0` (since it appears I no longer have permission, perhaps due to the recent migration). Then a thumbs up from him or @ldgauthier and I think this is good to go in!

**Review by ldgauthier — APPROVED** [review]
All the comparisons look great and I am confident in David's CNN->NV update plan -- let's do it!

**Comment by droazen** [comment]
Woohoo, thank you @samuelklee !!

**Comment by matthdsm** [comment]
@droazen, do you think this warrants a new point release? That way we can finally fix the `gatk-gcnvkernel` recipe over at bioconda and make the conda recipe useable again 😄

**Comment by droazen** [comment]
@matthdsm Yes definitely -- there will be another release fairly soon to get this out. Before we can release, though, we do need to merge a couple of PRs that have been waiting on this change (in particular, a replacement tool for CNNScoreVariants that uses PyTorch). We're currently targeting the late July / early August timeframe for the next release.

Are you the maintainer of the GATK bioconda recipes, by the way? Let us know if there's anything else we can do in the upcoming release to fix bioconda-related issues!

**Comment by matthdsm** [comment]
I'm _a_ bioconda maintainer, one of many, but I've got a vested interest in a functional gatk recipe 😅 
At the moment, we're unable to get the latest version of the GATK to build because of the requirements for the gcnvkernel.
A new version with the changes above would fix most if not all of the issues we're currently seeing.

**Comment by matthdsm** [comment]
Hi @droazen, 
Any updates on the timeframe for this new release? We're eagerly waiting for the next version so we can start updating everything on our side!

**Pull request merged**
